sdk/src/main/scala/com/horizen/SidechainState.scala

```diff
@@ -177,6 +178,47 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage,
     if (allInputBoxesIds.size != allInputBoxesIds.toSet.size) {
       throw new IllegalArgumentException(s"Block ${mod.id} contains duplicated input boxes to open")
     }
+
+    if (backwardTransferLimitEnabled)
+      checkWithdrawalBoxesAllowed(mod)
+  }
+
+  private def checkWithdrawalBoxesAllowed(mod: SidechainBlock): Unit = {
+    val alreadyMinedWBs = getAlreadyMinedWithdrawalRequestBoxesInCurrentEpoch
+    val mainchainBlockReferenceInBlock = mod.mainchainBlockReferencesData.size
+    val allowedWBs = getAllowedWithdrawalRequestBoxes(mainchainBlockReferenceInBlock)
+    var blockWBs = alreadyMinedWBs
+    mod.transactions.foreach(tx => {
+      blockWBs += tx.newBoxes().asScala.count(box => box.isInstanceOf[WithdrawalRequestBox])
+      if (blockWBs > allowedWBs) {
+        throw new IllegalStateException(s"Exceeded the maximum number of WithdrawalBoxes allowed!")
+      }
+    })
+  }
+
+  private def getOrElseWithdrawalEpochInfo(): WithdrawalEpochInfo = {
+    stateStorage.getWithdrawalEpochInfo.getOrElse(WithdrawalEpochInfo(0,0))
+  }
+
+  private def getAllowedWithdrawalRequestBoxesPerBlock: Int = {
+    params.maxWBsAllowed / (params.withdrawalEpochLength - 1)
+  }
+
+  def getAlreadyMinedWithdrawalRequestBoxesInCurrentEpoch: Int = {
+    stateStorage.getWithdrawalRequests(getOrElseWithdrawalEpochInfo().epoch).size
+  }
+
+  def getAllowedWithdrawalRequestBoxes(numberOfMainchainBlockReferenceInBlock: Int): Int = {
```

Comment:
The formula shouldn't be:
 open slots number =`3999 * (current_block_of_the_epoch + num_of_mc_block_ref_data_in_current_block) / (we_length - 1)` ?

Reply:
yes, it seems fine.
And moreover let's keep do then Min(3999, open_slots_number), in case in the future the last block of the epoch will be able to keep txs (to prevent the result > 3999)